README
NAME
    Crypt::SRP - Secure Remote Protocol (SRP6a)

SYNOPSIS
    Example 1 - SRP login handshake:

     ###CLIENT###
     my $I = '...'; # login entered by user
     my $P = '...'; # password entered by user
     my $cli = Crypt::SRP->new('RFC5054-1024bit', 'SHA1');
     my ($A, $a) = $cli->client_compute_A;

     #  request[1] to server:  ---> /auth/srp_step1 ($I, $A) --->

                               ###SERVER###
                               my %USERS;  # sort of "user database"
                               my %TOKENS; # sort of temporary "token database"
                               my $v = $USERS{$I}->{verifier};
                               my $s = $USERS{$I}->{salt};
                               my $srv = Crypt::SRP->new('RFC5054-1024bit', 'SHA1');
                               return unless $srv->server_verify_A($A);
                               $srv->server_init($I, $v, $s);
                               my ($B, $b) = $srv->server_compute_B;
                               my $token = $srv->random_bytes(32);
                               $TOKENS{$token} = [$I, $A, $B, $b];

     #  response[1] from server:  <--- ($B, $s, $token) <---

     ###CLIENT###
     return unless $cli->client_verify_B($B);
     $cli->client_init($I, $P, $s);
     my $M1 = $cli->client_compute_M1;

     #  request[2] to server:  ---> /auth/srp_step2 ($M1, $token) --->

                               ###SERVER###
                               my $M2 = '';
                               return unless $M1 && $token && $TOKENS{$token};
                               my ($I, $A, $B, $b) = @{delete $TOKENS{$token}};
                               return unless $I && $A && $B && $b && $USERS{$I};
                               my $s = $USERS{$I}->{salt};
                               my $v = $USERS{$I}->{verifier};
                               return unless $s && $v;
                               my $srv = Crypt::SRP->new('RFC5054-1024bit', 'SHA1');
                               $srv->server_init($I, $v, $s, $A, $B, $b);
                               return unless $srv->server_verify_M1($M1);
                               $M2 = $srv->server_compute_M2;
                               my $K = $srv->get_secret_K; # shared secret

     #  response[2] from server:  <--- ($M2) <---

     ###CLIENT###
     if ($M2 && $cli->client_verify_M2($M2)) {
       my $K = $srv->get_secret_K; # shared secret
       print "SUCCESS";
     }
     else {
       print "ERROR";
     }

    Example 2 - creating a new user and his/her password verifier:

     ###CLIENT###
     my $I = '...'; # login entered by user
     my $P = '...'; # password entered by user
     my $cli = Crypt::SRP->new('RFC5054-1024bit', 'SHA1');
     my ($v, $s) = $cli->compute_verifier_and_salt($I, $P);

     #  request to server:  ---> /auth/create_user [$I, $s, $v] --->

                               ###SERVER###
                               my %USERS;  # sort of "user database"
                               die "user already exists" unless $USERS{$I};
                               $USERS{$I}->{salt} = $s;
                               $USERS{$I}->{verifier} = $v;

    Working sample implementation of SRP authentication on client and server
    side is available in "examples" subdirectory: srp_server.pl
    <https://metacpan.org/source/MIK/Crypt-SRP-0.012/examples/srp_server.pl>
    , srp_client.pl
    <https://metacpan.org/source/MIK/Crypt-SRP-0.012/examples/srp_client.pl>
    .

DESCRIPTION
    More info about SRP protocol:

    *   <http://srp.stanford.edu/design.html>

    *   <http://en.wikipedia.org/wiki/Secure_Remote_Password_protocol>

    *   <http://tools.ietf.org/html/rfc5054>

    This module implements SRP version 6a.

    IMPORTANT: This module performs some big integer arithmetic via
    Math::BigInt. From performance reasons it is recommended to install
    Math::BigInt::GMP.

    IMPORTANT: This module needs a cryptographically strong random number
    generator. It tries to use one of the following:

    *   Crypt::PRNG - random_bytes()

    *   Crypt::OpenSSL::Random - random_bytes()

    *   Net::SSLeay - RAND_bytes()

    *   Crypt::Random - makerandom_octet()

    *   Bytes::Random::Secure - random_bytes()

    *   As an unsecure fallback it uses built-in rand()

METHODS
    Login and password ($I, $P) can be ASCII strings (without utf8 flag) or
    raw octets. If you want special characters in login and/or password then
    you have to encode them from Perl's internal from like this: "$I =
    encode('utf8', $I)" or "$P = encode('utf8', $P)"

    All SRP related variables ($s, $v, $A, $a, $B, $b, $M1, $M2, $S, $K) are
    by defaults raw octets (no BigInts, no strings with utf8 flag). However
    if you set new's optional parameter $format to 'hex', 'base64' or
    'base64url' SRP related input parameters (not $I or $P) are expected in
    given encoding and return values are converted into the same encoding as
    well.

    *   new

         my $srp = Crypt::SRP->new();
         #or
         my $srp = Crypt::SRP->new($group, $hash, $format, $interleaved, $default_salt_len);
         # $group ... (optional, DEFAULT='RFC5054-2048bit')
         #            'RFC5054-1024bit' or 'RFC5054-1536bit' or 'RFC5054-2048bit' or
         #            'RFC5054-3072bit' or 'RFC5054-4096bit' or 'RFC5054-6144bit' or
         #            'RFC5054-8192bit' see rfc5054 (appendix A)
         # $hash  ... (optional, DEFAULT='SHA256')
         #            'SHA1' or 'SHA256' or 'SHA384' or 'SHA512'
         # $format ... (optional, DEFAULT='raw')
         #             'raw' or 'hex' or 'base64' or 'base64url'
         # $interleaved ... (optional, DEFAULT=0) indicates whether the final shared
         #                  secret K will be computed as SHAx(S) or SHAx_Interleaved(S)
         #                  see rfc2945 (3.1 Interleaved SHA)
         # $default_salt_len ... (optional, DEFAULT=32)
         #                        default length (in bytes) for generated salt

    *   reset

         $srp->reset();
         #or
         $srp->reset($group, $hash, $format, $interleaved, $default_salt_len);  # see new()

         # returns $srp (itself)

    *   dump

         my $serialized_state = $srp->dump();

    *   load

         $srp->load($serialized_state);

    *   compute_verifier

         my $v = $srp->compute_verifier($I, $P, $s);

    *   compute_verifier_and_salt

         my ($v, $s) = $srp->compute_verifier_and_salt($I, $P);
         #or
         my ($v, $s) = $srp->compute_verifier_and_salt($I, $P, $s_len);

    *   client_init

         $srp->client_init($I, $P, $s, $B);

         # returns $srp (itself)

    *   client_compute_A

         my ($A, $a) = $srp->client_compute_A();
         #or
         my ($A, $a) = $srp->client_compute_A($a_len);

    *   client_compute_M1

         my $M1 = $srp->client_compute_M1($B);

    *   client_verify_M2

         my $valid = $srp->client_verify_M2($M2);

    *   client_verify_B

         my $valid = client_verify_B($B);

    *   server_init

         $srp->server_init($I, $v, $s);
         #or
         $srp->server_init($I, $v, $s, $A, $B, $b);
         # returns $srp (itself)

    *   server_compute_B

         my ($B, $b) = $srp->server_compute_B();
         #or
         my ($B, $b) = $srp->server_compute_B($b_len);

    *   server_fake_B_s

         my ($B, $s) = $srp->server_fake_B_s($I);
         #or
         my ($B, $s) = $srp->server_fake_B_s($I, $nonce);
         #or
         my ($B, $s) = $srp->server_fake_B_s($I, $nonce, $s_len);

    *   server_verify_M1

         my $valid = $srp->server_verify_M1($M1);

    *   server_compute_M2

         my $M2 = $srp->server_compute_M2();

    *   server_verify_A

         my $valid = server_verify_A($A);

    *   get_secret_S

         my $S = $srp->get_secret_S();
         #or
         my $S = $srp->get_secret_S($format);
         # $format can me 'raw' or 'hex' or 'base64' or 'base64url'

    *   get_secret_K

         my $K = $srp->get_secret_K();
         #or
         my $K = $srp->get_secret_K($format);
         # $format can me 'raw' or 'hex' or 'base64' or 'base64url'

    *   random_bytes

         my $rand = $srp->random_bytes();  # $rand formated according to $format passed to new()
         #or
         my $rand = $srp->random_bytes($len);

         my $rand = Crypt::SRP->random_bytes();  # $rand always raw bytes
         #or
         my $rand = Crypt::SRP->random_bytes($len);

LICENSE
    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

COPYRIGHT
    Copyright (c) 2012 DCIT, a.s. <http://www.dcit.cz> / Karel Miko



Changes
Changes for Crypt::SRP

0.016   2016/05/10
        - requires Math::BigInt-1.88 or newer

0.015   2016/04/13
        - minimizing dependencies, now only CryptX-0.030 + Math::BigInt-1.36

0.014   2014/10/07
        - fix compute_verifier

0.013   2014/01/17
        - optimized _calc_S_client

0.012   2013/09/25
        - documentation tweaks & cosmetics
        - URL to github repo

0.010   2013/06/17
        - adding README
        - removing utf8::downgrade workaround

0.009   2013/06/17
        - fix: hex padding when hex string has odd length

0.008   2013/06/07
        - calc_K fix (was broken for S with leading zeros)

0.007   2013/05/19
        - examples/srp_server.pl compatible with Mojoliciuos 4
        - use Crypt::PRNG if available
        - get_secret_S + get_secret_K accept optional $format parameter

0.006   2013/04/11
        - ($s, $v)=compute_verifier_and_salt
          >>> ($v,$s)=compute_verifier_and_salt
        - new($group, $hash, $interleaved, $format, $default_salt_len)
          >>> new($group, $hash, $format, $interleaved, $default_salt_len)
        - added server_client.t

0.005   2013/04/10
        - added reset(), dump(), load(), server_fake_B_s()
        - server_init() and client_init() now return $self
        - removed validate_A_or_B
        - added server_verify_A() + client_verify_B()
        - we support $self->random_bytes and Crypt::SRP->random_bytes
        - new's param $format + $default_salt_len

0.004   2013/04/08
        - fixed bug in _calc_K
        - new's param $interleaved
        - documentation fixes

0.003   2013/04/05
        - no dependency on Digest::SHA1
        - compatibility with Math::BigInt 1.991
        - more reliable sample implementation of client/server
        - ready for CPAN

0.002   2013/04/05
        - sample srp_client.pl + srp_server.pl (+ necessary changes)

0.001   2013/04/04
        - initial version


examples/srp_client.pl
#!/usr/bin/perl

# Copyright (c) 2013 DCIT, a.s. [http://www.dcit.cz] - Miko

use strict;
use warnings;
use Mojo::UserAgent;
use Crypt::SRP;

my $base_url = 'http://127.0.0.1:3000';
my $ua = Mojo::UserAgent->new;
#$ua->http_proxy('http://127.0.0.1:8081');

my $fmt = 'hex'; # all SRP related parameters are automatically converted from/to hex

my @test_set = ( ['alice', 'password123'] );
push @test_set, ["user$_", "secret$_"] for (1..3);

for (@test_set) {
  my $I = $_->[0];
  my $P = $_->[1];

  warn "#####################\n";
  warn "# [login=$I] SRP step1\n";

  my $cli = Crypt::SRP->new('RFC5054-1024bit', 'SHA1', 0, $fmt);
  #$cli->{predefined_a} = Math::BigInt->from_hex('60975527035CF2AD1989806F0407210BC81EDC04E2762A56AFD529DDDA2D4393'); #DEBUG-ONLY
  my ($A, $a) = $cli->client_compute_A(32);
  warn "I = $I\n";
  warn "P = $P\n";
  warn "A = ", substr($A,0,30), "..\n";
  warn "a = ", substr($a,0,30), "..\n";
  warn ">> gonna send request 1\n\n";
  my $tx1 = $ua->post("$base_url/auth/srp_step1" => json => {I=>$I, A=>$A});
  $tx1->res->json or warn("invalid response") and next;

  warn "# [login=$I] SRP step2\n";

  my $s = $tx1->res->json->{s};
  my $B = $tx1->res->json->{B};  
  my $token = $tx1->res->json->{token};
  $cli->client_verify_B($B) or warn("invalid B") and next;
  $cli->client_init($I, $P, $s);
  my $M1 = $cli->client_compute_M1();
  warn "token = $token\n";
  warn "s = $s\n";
  warn "B = ", substr($B,0,30), "..\n";
  warn "M1= $M1\n";
  warn ">> gonna send request 2\n\n";
  my $tx2 = $ua->post("$base_url/auth/srp_step2" => json => {M1=>$M1, token=>$token});
  $tx2->res->json or warn("invalid response") and next;

  warn "# [login=$I] SRP result\n";

  my $M2 = $tx2->res->json->{M2};
  warn "M2= $M2\n" if $M2;
  if ($M2 && $cli->client_verify_M2($M2)) {
     my $K = $cli->get_secret_K(); # shared secret
     warn "[SUCCESS] K = $K\n\n";
   }
   else {
     warn "[ERROR]\n\n";
   }
 }

examples/srp_server.pl
#!/usr/bin/perl

# requires Mojolicious 3.93+
# start like this: ./srp_server.pl daemon

# Copyright (c) 2013 DCIT, a.s. [http://www.dcit.cz] - Miko

use strict;
use warnings;
use Mojolicious::Lite;
use Crypt::SRP;

my %USERS;  # sort of "user database"
my %TOKENS; # sort of temporary "token database"
my $fmt = 'hex'; # all SRP related parameters are automatically converted from/to hex
my $dump;

%USERS = (
  alice => { # password P = "password123"
    salt     => 'beb25379d1a8581eb5a727673a2441ee',
    verifier => '7e273de8696ffc4f4e337d05b4b375beb0dde1569e8fa00a9886d8129bada1f1'.
                '822223ca1a605b530e379ba4729fdc59f105b4787e5186f5c671085a1447b52a'.
                '48cf1970b4fb6f8400bbf4cebfbb168152e08ab5ea53d15c1aff87b2b9da6e04'.
                'e058ad51cc72bfc9033b564e26480d78e955a5e29e7ab245db2be315e2099afb',
  }
);

my $cli = Crypt::SRP->new('RFC5054-1024bit', 'SHA1', $fmt);
for (1..3) {
  my $I = "user$_";
  my $P = "secret$_";
  my ($v, $s) = $cli->compute_verifier_and_salt($I, $P, 32);
  $USERS{$I} = { salt=>$s, verifier=>$v };
}
for (sort keys %USERS) {
  app->log->info(" [$_] s=$USERS{$_}{salt} v=" . substr($USERS{$_}{verifier},0,10) . "..");
}
  
post '/auth/srp_step1' => sub {
    my $self = shift;
    my $I = $self->req->json->{I};
    my $A = $self->req->json->{A};
    return $self->render_json({status=>'invalid'}) unless $I && $A;
    my $srv = Crypt::SRP->new('RFC5054-1024bit', 'SHA1', $fmt);
    return $self->render_json({status=>'invalid'}) unless $srv->server_verify_A($A);
    my $token = $srv->random_bytes(8);
    if ($USERS{$I} && $USERS{$I}->{salt} && $USERS{$I}->{verifier}) {
      # user exists
      my ($s, $v) = ($USERS{$I}->{salt}, $USERS{$I}->{verifier});
      $srv->server_init($I, $v, $s);
      #$srv->{predefined_b} = Math::BigInt->from_hex('E487CB59D31AC550471E81F00F6928E01DDA08E974A004F49E61F5D105284D20'); #DEBUG-ONLY
      my ($B, $b) = $srv->server_compute_B(32);
      $self->app->log->info("I = $I");
      $self->app->log->info("v = ". substr($v,0,30). "..");
      $self->app->log->info("B = ". substr($B,0,30). "..");
      $self->app->log->info("b = ". substr($b,0,30). "..");
      $TOKENS{$token} = $srv->dump;
      $self->app->log->info("storing state len=" . length($TOKENS{$token}));
      return $self->render_json({B=>$B, s=>$s, token=>$token});
    }
    else {
      # fake response for no-nexisting user
      $dump = undef;
      my ($B, $s) = $srv->server_fake_B_s($I);
      return $self->render_json({B=>$B, s=>$s, token=>$token});
    }
  };

post '/auth/srp_step2' => sub {
    my $self = shift;
    my $M1 = $self->req->json->{M1};
    my $token = $self->req->json->{token};
    $self->app->log->info("token = $token");
    $self->app->log->info("M1 = $M1");
    return $self->render_json({status=>'error'}) unless $M1 && $token && $TOKENS{$token};
    
    my $srv = Crypt::SRP->new->load($TOKENS{$token}); #restore state

    if (!$srv->server_verify_M1($M1)) {
      $self->app->log->info("server_verify_M1 FAILED!");
      return $self->render_json({status=>'error'});
    }
    my $M2 = $srv->server_compute_M2();
    my $K = $srv->get_secret_K();
    $self->app->log->info("M2 = $M2");
    $self->app->log->info("[SUCCESS] K = $K");
    return $self->render_json({M2=>$M2});
  };

app->start;
